src/middlewared/middlewared/plugins/directoryservices_/ipa_join_mixin.py

```diff
@@ -0,0 +1,419 @@
+        for etc_file in DSType.IPA.etc_files:
+            self.middleware.call_sync('etc.generate', etc_file)
+
+        self.middleware.call_sync('service.stop', 'sssd')
+        self.middleware.call_sync('service.start', 'sssd', {'silent': False})
+        self.middleware.call_sync('kerberos.start')
+
+    def _ipa_insert_keytab(self, service: ipa_constants.IpaConfigName, keytab_data: str) -> None:
+        """ Insert a keytab into the TrueNAS config (replacing existing) """
+        if service is ipa_constants.IpaConfigName.IPA_CACERT:
+            raise ValueError('Not a keytab file')
+
+        kt_name = service.value
+        if kt_entry := self.middleware.call_sync('kerberos.keytab.query', [
+            ['name', '=', kt_name]
+        ]):
+            self.middleware.call_sync(
+                'datastore.update', 'directoryservice.kerberoskeytab',
+                kt_entry[0]['id'],
+                {'keytab_name': kt_name, 'keytab_file': keytab_data}
+            )
+        else:
+            self.middleware.call_sync(
+                'datastore.insert', 'directoryservice.kerberoskeytab',
+                {'keytab_name': kt_name, 'keytab_file': keytab_data}
+            )
+
+    def _ipa_grant_privileges(self) -> None:
+        """ Grant domain admins ability to manage TrueNAS """
+
+        ipa_config = self.middleware.call_sync('ldap.ipa_config', ldap_config)
+
+        existing_privileges = self.middleware.call_sync(
+            'privilege.query',
+            [["name", "=", ipa_config['domain'].upper()]]
+        )
+
+        if existing_privileges:
+            return
+
+        try:
+            admins_grp = self.middleware.call_sync('group.get_group_obj', {
+                'groupname': 'admins',
+                'sid_info': True
+            })
+        except Exception:
+            self.logger.debug(
+                'Failed to look up admins group for IPA domain. API access for admin '
+                'accounts will have to be manually configured', exc_info=True
+            )
+            return
+
+        match admins_grp['source']:
+            case 'LDAP':
+                pass
+            case 'LOCAL':
+                self.logger.warning(
+                    'Local "admins" group collides with name of group provided '
+                    'by IPA domain, which prevents the IPA group from being '
+                    'automatically granted API access.'
+                )
+                return
+            case _:
+                self.logger.warning(
+                    '%s: unexpected source for "admins" group, which prevents '
+                    'the IPA group from being automatically granted API access.',
+                    admins_grp['source']
+                )
+                return
+
+        try:
+            self.middleware.call_sync('privilege.create', {
+                'name': ipa_config['domain'].upper(),
+                'ds_groups': [admins_group['gid']],
```

Comment:
flake8: `admins_group` undefined, presumably should be `admins_grp`?

Reply:
Yeah. flubbed.